### PR TITLE
Committed a time series plotting script

### PR DESCRIPTION
**Pull request by yixinmao**
@orianac Code review request - you can just focus on the structure of the script/functions!

### CODE REVIEW

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import numpy as np
+import datetime as dt
+import matplotlib.pyplot as plt
+import sys
+import pandas as pd
+import argparse
+import my_functions
+
+parser = argparse.ArgumentParser()
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py:14]: Maybe instead of reading arguments via command line arguments, you could try using a configuration file. I started using the read_config function that Joe wrote as part of tonic in some of my scripts. The function (see 
https://github.com/jhamman/tonic/blob/master/tonic/io.py) uses the SafeConfigParser and OrderedDict modules. It reads templates like this one https://github.com/orianac/BPA_304/blob/feature/sensitivity/tools/sensitivity/run_prms_sensitivity.cfg.template

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# Parameter setting
+#========================================================
+usgs_site_code = args.USGS_code # '03571850'
+usgs_site_name = args.USGS_name # 'Tennessee River at Lower Pittsburgh'
+
+hist_route_path = args.hist_route_path # historical run. Lohmann routing output. year; month; day; flow(cfs)
+future_route_paths = args.future_route_paths # future run. Lohmann routing output. year; month; day; flow(cfs); ['path1', 'path2']
+future_route_names = args.future_route_names # future run. year; month; day; flow(cfs); ['Scenario1', 'Scenario2']
+
+output_plot_basename = '../output/Tennessee_' + usgs_site_code # output plot path basename (suffix will be added to different plots)
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: The hard coding here could be removed by incorporating the variables into a config file.

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+plot_future_end_date = dt.datetime(2099, 9, 30)  # end date shown on the plot for future run
+
+time_locator = ('year', 20)  # time locator on the plot; 'year' for year; 'month' for month. e.g., ('month', 3) for plot one tick every 3 months
+
+#-------------------------------------------------
+
+nfuture = len(future_route_paths)  # number of future scenarios
+
+hist_style = 'k-'
+style_options = ['b--', 'r--', 'g--', 'm--', 'y--']
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: I like this dynamic way of making a style list because you'll be able to easily plot a different number of scenarios. But what happens if you have more than 5 future scenarios?

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+list_s_future = []  # a list contains pd.Series of data of each futre scenario
+for i in range(nfuture):
+	s_future = my_functions.read_Lohmann_route_daily_output(future_route_paths[i])
+	list_s_future.append(s_future)
+
+#========================================================
+# Select data to be plotted
+#========================================================
+s_hist_to_plot = my_functions.select_time_range(s_hist, plot_hist_start_date, plot_hist_end_date)
+list_s_future_to_plot = []
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: Instead of appending the series to a list, it seems like you could just loop through and add them to a pandas dataframe. That way you could easily truncate all timeseries uniformly instead of looping through and using the select_time_range function on every series independently as in line 68.

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+for i in range(nfuture):
+	s_future_to_plot = my_functions.select_time_range(list_s_future[i], plot_future_start_date, plot_future_end_date)
+	list_s_future_to_plot.append(s_future_to_plot)
+
+#========================================================
+# plot
+#========================================================
+model_info = 'VIC runoff, historical -  1/8 deg, Maurer forcing, Andy Wood setup\n          VIC runoff, climate change scenarios - 1/8 deg, Reclamation archive (GCM forcing)\n          Route: Lohmann, Wu flowdir, Andy Wood setup'
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: Adding this part to the config file would also be great. Also, it might be nice to include some more details about author/date. I wrote a little datestamp() function that I've been attaching to everything, including the output figure path so that every figure is datestamped both on the image itself and in the pathname. This way you don't overwrite figures the next day and you also have a record of them in the image itself (useful for presentations!).

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+##============== plot daily data ===============#
+#fig = my_functions.plot_time_series(plot_date=True, \
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: After looking at the my_functions.plot_time_series (I can't see it in this diff so don't know how to comment on it, but can look at it in your develop repo):
In lines 459-463 when you do plt.plot_date and access the pandas series information, you could instead just do series.plot() or dataframe.plot() and it will handle the looping through of each series for you. It might be cleaner and have tests for corner cases.

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+#		list_s_data=[s_hist_to_plot]+list_s_future_to_plot, \
+#		list_style=[hist_style]+list_style_future, \
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: The list_s_data, list_style, and list_label items could all be included into one dictionary which could make it more user-friendly. And then you could even make a dictionary with all of your plotting stylistic options and pass it to matplotlib as **kwargs (this is something I've been trying to do but haven't gotten to work yet). Like in this link: http://stackoverflow.com/questions/334655/passing-a-dictionary-to-a-function-in-python-as-keyword-parameters

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+#		list_label=['historical']+future_route_names, \
+#		plot_start=plot_hist_start_date, plot_end=plot_future_end_date, \
+#		ylabel='Flow (cfs)', title='%s, %s' %(usgs_site_name, usgs_site_code), 
+#		fontsize=16, legend_loc='upper left', \
+#		time_locator=time_locator,\
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: The time locator piece wouldn't be necessary with pandas!

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+#		add_info_text=True, model_info=model_info, stats='daily, no stats')
+#
+#fig.savefig('%s.future.flow.daily.png' %output_plot_basename, format='png')
+#
+##============== plot monthly data ===============#
+#fig = my_functions.plot_monthly_data(list_s_data=[s_hist_to_plot]+list_s_future_to_plot, \
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: It seems like the big change (in response to your request about focusing on the structure of the script/functions) could be to try to combine all of your separate plotting functions into one plotting function with different temporal options. 

i.e.: You have a separate function for plot_daily_data, plot_monthly_data, and plot_seasonality_data. Perhaps you could make a single function for plotting time series and have an option within it that decides on what time frame you want to plot/calculate statistics etc. Then if you end up further developing the plotting functionality those improvements will automatically be applied no matter what kind of timeseries plot you're making.

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+#============== plot seasonality data ===============#
+fig = my_functions.plot_seasonality_data(list_s_data=[s_hist_to_plot]+list_s_future_to_plot, \
+		list_style=[hist_style]+list_style_future, \
+		list_label=['historical']+future_route_names, \
+		plot_start=1, plot_end=12, \
+		ylabel='Flow (cfs)', \
+		title='Monthly mean seasinality, %s, %s' %(usgs_site_name, usgs_site_code), 
+		fontsize=16, legend_loc=None, \
+		xtick_location=range(1,13), \
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: Since for the seasonal plots you'll always be having 12 months and the xtick_location part won't change, you could absorb that piece into the function.

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+		xtick_labels=['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Nov','Oct','Nov','Dec'], \
+		add_info_text=True, model_info=model_info, stats='Monthly mean; historical - WY%d-%d; climate change scenario - WY%d-%d' %(plot_hist_start_date.year+1, plot_hist_end_date.year, plot_future_start_date.year+1, plot_future_end_date.year))
+
+fig = plt.savefig('%s.future.flow.seas.png' %output_plot_basename, format='png')
+
+#============== plot annual mean flow ===============#
+fig = my_functions.plot_WY_mean_data(list_s_data=[s_hist_to_plot]+list_s_future_to_plot, \
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: your plot_WY_mean_data appears twice in the my_functions.py script! This could pose problems if you make developments to the first that will get overwritten by the second!

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		list_style=[hist_style]+list_style_future, \
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: I like the test you have in the plot_WY_mean_data function. More tests! (I don't ever add tests, so we can work on this together ;)

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+		list_label=['historical']+future_route_names, \
+		plot_start=plot_hist_start_date.year, plot_end=plot_future_end_date.year, \
+		ylabel='Flow (cfs)', title='Annual mean (WY), %s, %s' %(usgs_site_name, usgs_site_code),
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: Since the function is by definition a plot_WY_mean function, the stubs of "Annual mean (WY)" could be absorbed into the function.

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
+#!/usr/local/anaconda/bin/python
+
+# This script plots different routed flow results together
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py:3]: Maybe more description of contact information/dates (though maybe not because it's on github and thus tied to you?)

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+hist_style = 'k-'
+style_options = ['b--', 'r--', 'g--', 'm--', 'y--']
+list_style_future = []  # a list of plotting styles of future scenarios
+for i in range(nfuture):
+	list_style_future.append(style_options[i])
+
+#========================================================
+# Load data
+#========================================================
+s_hist = my_functions.read_Lohmann_route_daily_output(hist_route_path)
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: In this function you load data with np.loadtxt then convert it to pandas series. I seem to remember reading somewhere that read_csv is faster (and then you also wouldn't have to go through the hoops of converting it to a pandas series yourself and risk having a bug along the way). Here's a comparison  with examples (albeit using np.genfromtxt) http://stackoverflow.com/questions/21486963/numpy-genfromtxt-slower-than-pandas-read-csv

File: VIC_RBM/model_run/result_analysis/scripts/my_functions.py [hunk 16]
```diff
@@ -265,17 +265,27 @@ def calc_monthly_data(data):
...
 		df: a [pd.DataFrame/Series] object, with index of time
-		by: string of group by, (select from 'year' or 'month')
+		by: string of group by, (select from 'year' or 'month' or 'WY')
 		stat: statistics to be calculated, (select from 'mean')
 		(e.g., if want to calculate monthly mean seasonality (12 values), by='month' and stat='mean')
 
 	Return:
 		A [dateframe/Series] object, with group as index (e.g. 1-12 for 'month')
+
+	Require:
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/my_functions.py:364]: I like the require piece- it makes it clearer for a user as to what they need to put in and keys us in to what's necessary.

File: VIC_RBM/model_run/result_analysis/scripts/my_functions.py [hunk 17]
```diff
@@ -527,5 +544,64 @@ def plot_seasonality_data(list_s_data, list_style, list_label, plot_start, plot_
...
+		plot_date_format
+		add_info_text_to_plot(fig, ax, model_info, stats)
+		plot_time_series
+		calc_ts_stats_by_group
+	'''
+
+	# Check if list_s_data, list_style and list_label have the same length
+	if len(list_s_data) !=len(list_style) or len(list_s_data)!=len(list_label):
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/my_functions.py:911]: Add more tests? Even if not for things that will break, but rather testing to see if your timeseries are of different lengths you could have it include a note on your figure describing which timeseries didn't have a complete record.

File: VIC_RBM/model_run/result_analysis/scripts/my_functions.py [hunk 18]
```diff
@@ -527,5 +544,64 @@ def plot_seasonality_data(list_s_data, list_style, list_label, plot_start, plot_
...
+		print 'Input list lengths are not the same!'
+		exit()
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/my_functions.py:913]: Or you could also provide a fall-back instead of having the function break. You could provide a default styles list or a labels list that says "EMPTY NAME" or something. Or construct it with a try/except statement to give a more user-friendly exit experience.

File: VIC_RBM/model_run/result_analysis/scripts/my_functions.py [hunk 19]
```diff
@@ -527,5 +544,64 @@ def plot_seasonality_data(list_s_data, list_style, list_label, plot_start, plot_
...
+	Require:
+		plot_date_format
+		add_info_text_to_plot(fig, ax, model_info, stats)
+		plot_time_series
+		calc_ts_stats_by_group
+	'''
+
+	# Check if list_s_data, list_style and list_label have the same length
+	if len(list_s_data) !=len(list_style) or len(list_s_data)!=len(list_label):
+		print 'Input list lengths are not the same!'
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/my_functions.py:912]: Also, this is a test that would be useful in any of the plotting functions you have, so you could either pull it out to happen at the beginning of the script entirely or you could collapse all of the plotting function such that the test is always read when plotting (no matter on what time frame).

File: VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py (hunk 2 above)
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/plot_flow_future.py]: In terms of the path names, I just saw from comments on my own pull request that this should be changed to using os.join()

File: VIC_RBM/model_run/result_analysis/scripts/my_functions.py [hunk 20]
```diff
@@ -265,17 +265,27 @@ def calc_monthly_data(data):
 #==============================================================
 #==============================================================
 
+def wateryear(calendar_date):
```
Comment (orianac) [inline VIC_RBM/model_run/result_analysis/scripts/my_functions.py:347]: This is probably in tonic?

**Comment by bartnijssen** [comment]
Only commenting on the pull request (PR) is not very useful in this case, since we have never reviewed the original file. It'd be easier to comment on the whole file in `develop`
